src/Val.h

```diff
@@ -1224,6 +1247,27 @@ class RecordVal final : public Val, public notifier::detail::Modifiable {
 	IntrusivePtr<T> GetFieldOrDefault(const char* field) const
 		{ return cast_intrusive<T>(GetField(field)); }
 
+	// The following return the given field converted to a particular
+	// underlying value.  We provide these to enable efficient
+	// access to record fields (without requiring an intermediary Val)
+	// if we change the underlying representation of records.
+#define GET_FIELD_AS(ctype, name ) \
```

Comment:
No new macros, please. :-)

This should be doable through templated methods.

Reply:
I'll need guidance, then.  The part I don't see is how to select the right `AsXYZ()` converter, given those aren't templated.

Reply:
This is probably another instance of my question above: why the preference for the `AsX` macros()?

I'd turn this into either `GetField<AddrVal>()->Get()` or `GetField<AddrVal>()` (the latter adding the `Get()` internally). Or actually, maybe even both versions to get to either the `Val` or the underlying value:

```
class RecordVal {
....
template<typename T>
auto GetFieldVal(int field) const { return dynamic_vast<T*>(GetField(i)); }

template<typename T>
auto GetField(int field) const { return GetFieldVal<T>(i)->Get(); }

}
```

Reply:
One question re going this route: in the future when records can be vectors of `ZAMValUnion` (or whatever name we wind up using), is the templating machinery powerful enough to allow picking out the right field from the union for a given `T` type?  I imagine it is for a `std::variant` representation, but I'd rather defer the decision on union-vs-variant and not presuppose it by what's done here.

Reply:
We haven't fully discussed how the internal representation is going to look like eventually (I was thinking we'd copy it out into a ZAMValUnion, not storing a ZAMValUnion. We can see, but I hink this is less a question of the templating and more of the cast: as long as we can get a ValPtr out of the union/variant, we can cast it to our destination type (hmm... I guess that's goes back to the question if the union can just store a ValPtr rather than the individual types).

Reply:
I'm not following here.  One of the benefits of ZAMValUnion is that atomic types like `int` / `port` / `interval` are represented with no overhead.  That's lost if access to them needs to go through ValPtr's.

Reply:
Ok, nevermind. Agree that that is one of the main benefits, I might just be getting confused on some parts I hadn't fully followed through yet.

Reply:
I unresolved this one because the original point remains unaddressed: I'd really really like to avoid adding any new macros. Also, on further consideration, I do think the template machinery could be taught to pick out the right field even with a union (one approach would be a hardcoded type-based switch; maybe there's a nicer way too). So I'll keep my proposal of the two templated `GetField` methods. (In the end, there's not much really that macros can do that templates can't do (and better)).

Reply:
I agree with avoiding a new macro. The one issue with making this a template though is that the function names that `GET_FIELD_AS` generates don't really match the actual type being returned. We can't do something like `As<Enum>` because `Enum` isn't a type. On top of that, C++ doesn't have the ability to do switches on types (that's a C#-ism, I think). We could do a block of `constexpr if` tests I think, but I haven't tested that theory yet.

Reply:
Yeah, I think the `constexpr` route would work I think, but it doesn't seem great.

But why wouldn't the above template approach work? One wouldn't parameterize it with the return value but with the `Val`-type: `f->GetField<EnumVal>(i)`. That turns into some version of `dynamic_cast<EnumVal*>(GetField(i))->Get()`, which yields the integer. Each `Val`-derived class "knows" its underlying type, and `Get()` returns that.

That said, I'm wondering if we even need this right now. I realize this is another instance of "preparing for later", but I find this shortcut to provide little benefit on its own: writing `GetEnumField(5)` instead of `GetField(5)->AsEnum()` feels actually *less* intuitive to me. So we could also just leave this out for now and revisit once we get to whatever will need something else here.

Reply:
The point is to enable efficient access if/when we change the underlying representation of records.  In that case, using a generic `GetField` will require constuction of a `Val` only to immediately discard it.

Reply:
> But why wouldn't the above template approach work? One wouldn't parameterize it with the return value but with the Val-type: f->GetField<EnumVal>(i). That turns into some version of dynamic_cast<EnumVal*>(GetField(i))->Get(), which yields the integer. Each Val-derived class "knows" its underlying type, and Get() returns that.

My issue with that is that `GetField<EnumVal>` isn't terribly clear about what it's returning. Is it an `EnumVal`? Something else? Maybe `GetField` is just not a good name for it.

Reply:
But there's the more basic issue of wanting to enable access in the future without having to construct a `Val`.

Reply:
Are you confident that's actually an issue, not something the compiler would just optimize away?

Reply:
How can the optimizer possibly get rid of it, when it involves constructing and destructing not just a `Val` but also its underlying `Obj`, and the latter has logic about calling `PLUGIN_HOOK_VOID` etc?

Reply:
(Not to mention needing to understand calls to `val_mgr`)

Reply:
I guess I'm not understanding why it would need to construct a `Val` in the templated version. Even if the table stores a `ZAMValUnion` directly, it simply needs to implement a `Get()` method of some sort and the current templated version works. If it doesn't, we can adapt the templated method to deal with that.

Reply:
I guess the part I'm missing is just what the intermediary function returns.  From the above (the `GetField<AddrVal>()->Get()` example), it looks like it would return a `Val*`.

Reply:
That's what the bit of `std::invoke_result_t<decltype(&T::Get), T>` magic does. It sets the return value of the method to whatever the return value of the `Get()` method for the type passed in. For things like `CountVal`, that would just be the internal `double` value. For more complicated things like `PortVal`, it returns a `const PortVal*` because that's needed in order to grab the internal fields.

Reply:
Update: @timwoj and I sync'd up off line, as he/Robin thought, it'll be workable to use templates for this.